tmt/steps/provision/bootc.py

```diff
@@ -0,0 +1,241 @@
+        help="""
+             Add tmt dependencies to the supplied container image or image built
+             from the supplied Containerfile.
+             This will cause a derived image to be built from the supplied image.
+             """)
+
+    image_builder: str = field(
+        default=DEFAULT_IMAGE_BUILDER,
+        option=('--image-builder'),
+        metavar='IMAGEBUILDER',
+        help="""
+             The full repo:tag url of the bootc image builder image to use for
+             building the bootc disk image.
+             """)
+
+
+@tmt.steps.provides_method('bootc')
+class ProvisionBootc(tmt.steps.provision.ProvisionPlugin[BootcData]):
+    """
+    Provision a local virtual machine using a bootc container image
+
+    Minimal config which uses the Fedora bootc image:
+
+    .. code-block:: yaml
+
+        provision:
+            how: bootc
+            containerimage: quay.io/fedora/fedora-bootc:40
+
+    Here's a config example using a containerfile:
+
+    .. code-block:: yaml
+
+        provision:
+            how: bootc
+            containerfile: "./my-custom-image.containerfile"
+            containerfile-workdir: .
+            image_builder: quay.io/centos-bootc/bootc-image-builder:stream9
+            disk: 100
+
+    Another config example using an image that includes tmt dependencies:
+
+    .. code-block:: yaml
+
+        provision:
+            how: bootc
+            add_deps: false
+            containerimage: localhost/my-image-with-deps
+
+    This plugin is an extension of the virtual.testcloud plugin.
+    Essentially, it takes a container image as input, builds a
+    bootc disk image from the container image, then uses the virtual.testcloud
+    plugin to create a virtual machine using the bootc disk image.
+
+    The bootc disk creation requires running podman as root, this is typically
+    done by running the command in a rootful podman-machine. The podman-machine
+    also needs access to ``/var/tmp/tmt``. An example command to initialize the
+    machine:
+
+    .. code-block:: shell
+
+        podman machine init --rootful --disk-size 200 --memory 8192 \
+        --cpus 8 -v /var/tmp/tmt:/var/tmp/tmt -v $HOME:$HOME
+    """
+
+    _data_class = BootcData
+    _guest_class = GuestTestcloud
+    _guest = None
+    _id = str(uuid.uuid4())[:8]
+
+    def _get_id(self) -> str:
+        # FIXME: cast() - https://github.com/teemtee/tmt/issues/1372
+        parent = cast(tmt.steps.provision.Provision, self.parent)
+        assert parent.plan is not None
+        assert parent.plan.my_run is not None
+        assert parent.plan.my_run.unique_id is not None
+        return parent.plan.my_run.unique_id
+
+    def _build_derived_image(self, base_image: str) -> str:
+        """ Build a "derived" container image from the base image with tmt dependencies added """
+        if not self.workdir:
+            raise tmt.utils.ProvisionError(
+                "self.workdir must be defined")
```

Comment:
To an end user, this error message is not very helpful. IIUIC, `assert self.workdir is not None  # narrow type` should be good enough, it's a common pattern that `self.workdir` remains `Optional`.